pkg/cmd/pulumi/config_env_init.go

```diff
@@ -0,0 +1,333 @@
+// Copyright 2016-2023, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"text/template"
+
+	"github.com/charmbracelet/glamour"
+	"github.com/erikgeiser/promptkit/confirmation"
+	"github.com/pulumi/esc"
+	"github.com/pulumi/esc/eval"
+	"github.com/pulumi/pulumi/pkg/v3/backend"
+	"github.com/pulumi/pulumi/pkg/v3/backend/display"
+	"github.com/pulumi/pulumi/pkg/v3/backend/httpstate"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource/config"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/workspace"
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+)
+
+func newConfigEnvInitCmd(stackRef *string) *cobra.Command {
+	var envName string
+	var showSecrets bool
+	var keepConfig bool
+	var yes bool
+
+	cmd := &cobra.Command{
+		Use:   "init",
+		Short: "Creates an environment for a stack",
+		Long: "Creates an environment for a specific stack based on the stack's configuration values,\n" +
+			"then replaces the stack's configuration values with a reference to that environment.\n" +
+			"The environment will be created in the same organization as the stack.",
+		Args: cmdutil.NoArgs,
+		Run: cmdutil.RunFunc(func(cmd *cobra.Command, args []string) error {
+			if !yes && !cmdutil.Interactive() {
```

Comment:
Can we write these commands in the new struct style (see https://github.com/pulumi/pulumi/blob/master/pkg/cmd/pulumi/plugin_install.go#L77 for example) and write tests for them? 
Might be tricky cause some parts of the system are horribly unmockable but would be useful if they can be tested.

Reply:
Hah, we're using the same style in `esc` for likely the same reasons :)

Yeah, I'll see if these fit into that style. Primary blocker is likely to be the unmockability that you mention.

Reply:
> Hah, we're using the same style in esc for likely the same reasons :)

Yeh there's part of me that doesn't feel totally comfortable with it, feel like we're missing something by not running the cobra bits through tests as well, but it's the best style we've got so far.